src/query/parser/promql/types.go

```diff
@@ -145,23 +144,29 @@ func newScalarOperator(
 	expr *promql.NumberLiteral,
 	tagOpts models.TagOptions,
 ) (parser.Params, error) {
-	return scalar.NewScalarOp(
-		func(_ time.Time) float64 { return expr.Val },
-		scalar.ScalarType,
-		tagOpts,
-	)
+	return scalar.NewScalarOp(expr.Val, tagOpts)
+}
+
+func isTime(expr promql.Expr) bool {
+	return expr.String() == "time()"
 }
 
 // NewBinaryOperator creates a new binary operator based on the type.
 func NewBinaryOperator(expr *promql.BinaryExpr,
 	lhs, rhs parser.NodeID) (parser.Params, error) {
 	matching := promMatchingToM3(expr.VectorMatching)
+	// NB: Prometheus handles `time()` in a way that's not entirely compatible
+	// with the query engine; special handling is required here.
+	lTime, rTime := isTime(expr.LHS), isTime(expr.RHS)
+	if matching == nil {
```

Comment:
nit: Could we make `promMatchingToM3` return `*thing, bool` instead of using `nil` to specify it didn't convert the expr?  Just tends to be the pattern we use rather than checking whether thing is nil or not (which can lead to some people forgetting to nil check, and then nil ptr dereference).